lib/runtime/src/ops/transpose.cc

```diff
@@ -50,369 +52,549 @@ TransposeParams Transpose::get_params() const {
-                                                    READ_ONLY,
-                                                    EXCLUSIVE,
-                                                    inputs[0]->region));
-  launcher.add_field(0, FID_DATA);
-  launcher.add_region_requirement(RegionRequirement(outputs[0]->part,
-                                                    0 /*projection id*/,
-                                                    WRITE_ONLY,
-                                                    EXCLUSIVE,
-                                                    outputs[0]->region));
-  launcher.add_field(1, FID_DATA);
-  FutureMap fm = runtime->execute_index_space(ctx, launcher);
-  fm.wait_all_results();
-  set_opmeta_from_futuremap(ff, fm);
+static DeviceSpecific<TransposePerDeviceState>
+    init_task(Task const *task,
+              std::vector<PhysicalRegion> const &regions,
+              Context ctx,
+              Runtime *runtime) {
+  TaskArgumentAccessor acc(task, regions, ctx, runtime);
+  return init_task_impl(acc);
 }
 
-void Transpose::init_meta(TransposeMeta *m,
-                          Domain const &in_domain,
-                          Domain const &out_domain) const {
-  for (int i = 0; i < out_domain.get_dim(); i++) {
-    assert(out_domain.hi()[i] == in_domain.hi()[this->perm[i]]);
-    assert(out_domain.lo()[i] == in_domain.lo()[this->perm[i]]);
-  }
-  m->num_dim = out_domain.get_dim();
-  for (int i = 0; i < m->num_dim; i++) {
-    m->perm[i] = this->perm[i];
-  }
+template <>
+void register_task<TRANSPOSE_INIT_TASK_ID>();
+OpTaskSignature init(OpTaskType::INIT)
+
+    init.add_arg_slot<TransposeAttrs>(ATTRS);
+register_task(TRANSPOSE_INIT_TASK_ID, "Transpose::init", init, init_task);
+} // namespace FlexFlow
+
+OpTaskInvocation forward(TransposeAttrs const &) {
+  OpTaskBinding binding;
+
+  binding.bind_arg(PER_DEVICE_STATE,
+                   per_device_op_state<TransposePerDeviceState>());
+  binding.bind_arg(PROFILEING, profiling_settings());
+
+  bind.bind(INPUT, input_tensor());
+  bind.bind(OUTPUT, output_tensor());
+
+  return {TRANSPOSE_FWD_TASK_ID, binding};
 }
 
-PerDeviceOpState *
-    Transpose::init_task(Task const *task,
-                         std::vector<PhysicalRegion> const &regions,
-                         Context ctx,
-                         Runtime *runtime) {
-  assert(regions.size() == 2);
-  assert(task->regions.size() == 2);
-  Transpose const *transpose = (Transpose const *)task->args;
-  FFHandler handle = *((FFHandler const *)task->local_args);
+static optional<float> forward_task_impl(TaskArgumentAccessor const &acc) {
+  ProfilingSettings profiling = acc.get_argument<ProfilingSettings>(PROFILING);
+  auto per_device_state =
+      acc.get_per_device_state<TransposePerDeviceState>(PER_DEVICE_STATE);
+
+  auto input = acc.get_tensor<Permissions::RO>(INPUT);
+  auto output = acc.get_tensor<Permissions::WO>(OUTPUT);
+
+  // copy from the old code
+  Task *task = acc.task;
```

Comment:
this may have problem

```
  Task *task = acc.task;
  Context ctx = acc.ctx;
  Runtime *runtime = acc.runtime;
  Domain in_domain = runtime->get_index_space_domain(
      ctx, task->regions[0].region.get_index_space());
  Domain out_domain = runtime->get_index_space_domain(
      ctx, task->regions[1].region.get_index_space());

  for (int i = 0; i < m->num_dim; i++) {
    assert(out_domain.hi()[i] == in_domain.hi()[m->perm[i]]);
    assert(out_domain.lo()[i] == in_domain.lo()[m->perm[i]]);
  }
```